tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -0,0 +1,266 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <gtest/gtest.h>
+
+#include <executor.h>
+#include <expr_evaluator.h>
+#include <fusion.h>
+#include <fusion_segmenter.h>
+#include <ir/all_nodes.h>
+#include <ir/interface_nodes.h>
+#include <ir/utils.h>
+#include <mma_type.h>
+#include <ops/all_ops.h>
+#include <preseg_passes/allocation_order_inference.h>
+#include <preseg_passes/move_split_cat.h>
+#include <preseg_passes/optimization_pass.h>
+#include <scheduler/mma_utils.h>
+#include <scheduler/utils.h>
+#include <tests/cpp/multidevice.h>
+#include <tests/cpp/validator.h>
+
+namespace nvfuser {
+
+class DistributedTransformerTest
+    : public MultiDeviceTest,
+      public testing::WithParamInterface<std::tuple<bool, DataType>> {
+ protected:
+  DistributedTransformerTest()
+      : D(communicator_->size()),
+        optimization_guard_(false),
+        allocation_order_guard_(false) {
+    NVF_CHECK(E % H == 0);
+    NVF_CHECK(H % D == 0);
+    NVF_CHECK(E % D == 0);
+  }
+
+  void SetUp() {
+    MultiDeviceTest::SetUp();
+    if (!deviceMajorMinorCheck(8)) {
+      GTEST_SKIP() << "Distributed transformer tests require Ampere or newer";
+    }
+  }
+
+  hir::HostIrExecutorParams executor_params_{
+      .use_fusion_executor_cache = true,
+      .skip_auto_scheduling = false,
+      .cache_fusion_executor = false};
+  const int D;
+  const int B = 4;
+  const int E = 128;
+  const int H = 4;
+  const int S = 32;
+
+  // Temporary until https://github.com/NVIDIA/Fuser/issues/2563
+  at::Tensor shardTensor(at::Tensor tensor, int64_t axis, DeviceMesh& mesh) {
```

Comment:
Note: added, to make it easier to express sharding where the sharded axis is divisible by the mesh size. Also adds the superficial DID axis.

Reply:
This seems to overlap quite a bit with MultiDeviceTest::shardTensor. How about the following: 

```
class MultiDeviceTest {
  ...
  at::Tensor shardTensor(at::Tensor tensor, int64_t axis, const DeviceMesh& mesh);
  at::Tensor shardTensor(at::Tensor tensor, TensorView* tv) {
    return shardTensor(tensor, getShardedAxis(tv), tv->getDeviceMesh());
  }
  ...
};

Reply:
Got it. However, I [was proposing](https://github.com/NVIDIA/Fuser/pull/2574#discussion_r1675329805) the current shardTensor to be a thin wrapper of the new shardTensor. Would that work?

Reply:
ahhhh got it! Yes that's good suggestion